sdk/src/fungible_token.ts

```diff
@@ -0,0 +1,274 @@
+    meta: ProgramMetadata,
+    programId: HexString,
+    gasLimit: bigint
+  ) {
+    FungibleToken.validateMetadata(meta)
+    this.api = api
+    this.programId = programId
+    this.meta = meta
+    this.eventListener = eventListener
+    this.gasLimit = gasLimit
+  }
+  static validateMetadata(meta: ProgramMetadata) {
+    if (meta.types.init.input === null) {
+      throw new Error('Metadata does not contain init type')
+    }
+
+    if (meta.types.handle.input === null) {
+      throw new Error('Metadata does not contain handle input type')
+    }
+
+    if (meta.types.handle.output === null) {
+      throw new Error('Metadata does not contain handle output type')
+    }
+
+    if (getStateInput(meta) === null) {
+      throw new Error('Metadata does not contain state input type')
+    }
+
+    if (meta.getTypeIndexByName(MetaDataTypes.u128) === null) {
+      throw new Error('Metadata does not contain u128 type')
+    }
+
+    if (meta.getTypeIndexByName(MetaDataTypes.u64) === null) {
+      throw new Error('Metadata does not contain u64 type')
+    }
+
+    if (meta.getTypeIndexByName(MetaDataTypes.OptionU64) === null) {
+      throw new Error('Metadata does not contain Option<u64> type')
+    }
+
+    if (meta.getTypeIndexByName(MetaDataTypes.u8) === null) {
+      throw new Error('Metadata does not contain u8 type')
+    }
+  }
+  static async deploy(
+    api: GearApi,
+    eventListener: EventListener,
+    deployer: KeyringPair,
+    name: string = '',
+    symbol: string = '',
+    decimals: bigint = 0n
+  ) {
+    const { metadata, wasm } = await getDeploymentData('fungible_token')
+    const inputType = metadata.types.init.input!
+
+    const init = metadata.createType(inputType, {
+      name,
+      symbol,
+      decimals
+    })
+
+    const gas = await api.program.calculateGas.initUpload(
+      `0x${Uint8ArrayToHexStr(deployer.publicKey)}`,
+      wasm,
+      init.toHex(), // payload
+      0, // value
+      false // allow other panics
+    )
+
+    const program = {
+      code: wasm,
+      gasLimit: gas.min_limit.toNumber() * 2,
+      value: 0,
+      initPayload: init.toHex()
+    }
+
+    const { programId, extrinsic } = api.program.upload(program, metadata)
+
+    const event: Promise<ISubmittableResult> = new Promise(resolve => {
+      extrinsic.signAndSend(deployer, async result => {
+        if (result.isFinalized) {
+          resolve(result)
+        }
+      })
+    })
+
+    const res = await event
+
+    if (res.isError) {
+      throw new Error(res.dispatchError?.toString())
+    } else {
+      return new FungibleToken(api, eventListener, metadata, programId, FUNGIBLE_TOKEN_GAS_LIMIT)
+    }
+  }
+
+  async sendMessage(
+    user: KeyringPair,
+    payload: Codec,
+    inputType: number
+  ): Promise<FungibleTokenResponse> {
+    const message: MessageSendOptions = {
+      payload: payload.toU8a(),
+      gasLimit: this.gasLimit,
+      destination: `${this.programId}`
```

Comment:
Is  this **destination: this.programId** not the same as this: **destination: \`${this.programId}\`** but with easier syntax?